### ISSUE TITLE
Bringing Filter to 0.10.x

### ISSUE THREAD

**Issue by joaomdmoura** [rails-api/active_model_serializers#1058]
### **tl;dr**
Older versions of AMS had a `filter`feature that older users are used to. We would like to bring this to 0.10.x.

---
Older versions of AMS had a `filter`feature that older users are used to.
We would like to bring this to 0.10.x mainly for two reasons:

1. There is no easy way to filter attributes on 0.10.x so far
2. It will help ppl to update from 0.8 and 0.9

The idea is to provide a method named filter, which should return an array used to determine what attributes and associations should be included in the output. This is typically used to customize output based on current_user. For example:

```ruby
class PostSerializer < ActiveModel::Serializer
  attributes :id, :title, :body, :author

  def filter(keys)
    if scope.admin?
      keys
    else
      keys - [:author]
    end
  end
end
```

---
related to issues:
- https://github.com/rails-api/active_model_serializers/issues/459

related to PRs:
- https://github.com/rails-api/active_model_serializers/pull/460
- https://github.com/rails-api/active_model_serializers/pull/1266

**Comment by bf4** [comment]
This is also a feature of the JSON API 

- http://jsonapi.org/format/#fetching-filtering
- http://jsonapi.org/recommendations/#filtering

**Comment by joaomdmoura** [comment]
But in specific to JSON-API I think the `fields` implementation is covering it.

**Comment by bf4** [comment]
Not sure I understand. Fields is different.

- http://jsonapi.org/format/#fetching-sparse-fieldsets
- http://jsonapi.org/examples/

**Comment by joaomdmoura** [comment]
Yup, you're right I messed up.
It seems it's not necessary but it's a recommendation that we should aim for.

It seems to me that some cases are not so useful, taking into considerations the rails context we usually have, like:
`GET /comments?filter[post]=1`

But as I said,  is part of the convention and we should aim for that. :smile: 
Would be awesome to have `filter` back working out of the box with the filtering patter on json-api.

**Comment by BenMorganIO** [comment]
I was just working on this, this was my implementation:

```ruby
Spree::Api::BaseController.class_eval do
  protected

  def index(collection = [])
    render json: collection.where(filter_params).paginate(page_params), include: params[:include]
  end

  private

  def filter_params
    params.fetch(:filter, {}).permit(filter_attributes << :id).transform_values { |v| v.split(',') }
  end

  def filter_attributes
    serializer = "Spree::#{controller_name.titlecase.singularize}Serializer"
    serializer.constantize._attributes
  end
end
```

Its based on the controller and I think the JSON API's definition of a filter is to identify objects that have an id of `1,2,3,..,n+1` or what if you have a prices endpoint and you want to filter by currency? Use cases:

```
/api/prices?filter[currency]=CAD,USD
/api/countries?filter[iso3]=CAD,USA&include=states
```

The filtering should be more controller based since we're focusing more on content coming from a URI and deciphering what is and isn't secure to accept. The fields is more of trying to focus on minimizing the JSON payload; especially over making a whitelist that _doesn't_ include certain relationships (like not wanting 50,000 ids back over some requests, but not others).

**Comment by beauby** [comment]
Considering what the spec says:
> JSON API is agnostic about the strategies supported by a server. The filter query parameter can be used as the basis for any number of filtering strategies.

I think it is not AMS's responsibility to handle this parameter.

**Comment by beauby** [comment]
However, the 0.8.0 filter feature (which is unrelated to the JSONAPI filter) remains a useful tool that I believe should be available in 0.10.0.

**Comment by BenMorganIO** [comment]
@beauby considering how easy it is to do on ones own, maybe make another gem that just adds more conventional sugar.

**Comment by BenMorganIO** [comment]
The 0.8.0 "filter" feature might need to be renamed to "fields". I think that's how the JSON API spec uses the terms.

**Comment by beauby** [comment]
Thing is, I'm not even sure what AMS should do with the `filter` query param if it were to do anything. Surely, filtering a collection before rendering should be done at the DB level.

**Comment by beauby** [comment]
@BenMorganIO The 0.8.0 filter is not really the same as the JSONAPI fields, as the former allows for the dev to pick which attributes/relationships will be serialized depending on certain conditions (like "is the current user an admin?"), whereas the latter allows the client to specify which fields it wishes to receive.

**Comment by BenMorganIO** [comment]
There's an `include` helper, so I think there's room for an argument on why the `filter` and `fields` params should be a part of the AMS 0.10.

However, I would opt for AMS to be have a second gem that adds that sugar to ActionController.

**Comment by BenMorganIO** [comment]
@beauby I like how the filter can do that based on the current users role. I'm :+1: for that. I'd rather opt for something like a `public_attributes`, `user_attributes`, `admin_attributes`. I feel like that would would be a much more intuitive API.

**Comment by beauby** [comment]
Again, `include` is intended to be usable by the dev to specify which relationships they want to include in the response. As a commodity, AMS supports the JSONAPI format for the `include` query parameter, but it's not limited to that scenario.

I'm all up for having JSONAPI-related helpers somewhere (I think that'd be great). I just don't think they all belongs within AMS.

**Comment by BenMorganIO** [comment]
@beauby I'm all :+1: for a JSONAPI-related helper gem for AMS. I'd love to put my JSON API `filters` params as a helper somewhere.

**Comment by vasilakisfil** [comment]
I am totally against having the filtering logic inside the AMS classes. However there should definitely be an option for filtering attributes in a serializer, but outside the AMS class. This was possible in the 0.9.x with the only: option. Maybe relevant to #1141

**Comment by vasilakisfil** [comment]
I think I can provide you a gem that filters attributes of a resource in the db level for ActiveRecord and Mongoid using AR's select(*attributes) and Mongoid's only respectively. I would create that anyway next week so why not coordinating with you..

**Comment by BenMorganIO** [comment]
So, Spree's V2 API actually can filter attributes. What I want to do is focus it on `self` and support associations later on. See: https://github.com/spree-contrib/spree_api_v2/blob/master/app/controllers/spree/api/v2/base_controller.rb#L28

You can see it in action here: http://spree-contrib.github.io/spree_api_v2/#filtering

However, what I want it to be able to do is:

```shell
# current
curl "https://example.com/api/articles?filter[id]=2,3,4"

# hope
curl "https://example.com/api/articles?filter[self.id]=2,3,4"
curl "https://example.com/api/articles?filter[comments.id]=2,3,4"
```

**Comment by bf4** [comment]
@joaomdmoura Can be closed to due merge of https://github.com/rails-api/active_model_serializers/issues/1403 ?

Other related issues/prs:

- https://github.com/rails-api/active_model_serializers/issues/857
- https://github.com/rails-api/active_model_serializers/issues/1333
- https://github.com/rails-api/active_model_serializers/pull/1538

**Comment by joaomdmoura** [comment]
Yup! #1403 seems to solve it in a better and more elegant way :grin: I'm closing it!
